---
title: "fetch_multi_align"
author: "Gokberk Alagoz"
date: "September 28, 2021"
output: html_document
---

```{r setup, include=FALSE}
library(stringr)
library(httr)
library(jsonlite)
library(xml2)
library(ape)
```

## R Markdown

```{r set_paths}
lead_snps = read.table("/data/workspaces/lag/workspaces/lg-ukbiobank/projects/globularity/evolution/results/VEP/globularity_lead_snps_VEP.txt", header = T)

# remove SNPs that are mapped to patches
lead_snps = lead_snps[!str_detect(lead_snps$Location, "CHR"), ]

outDir = "/data/workspaces/lag/workspaces/lg-ukbiobank/projects/globularity/evolution/results/phylogenetic_context/multiple_seq_alignments/"

```

## SNP list liftover (GRCh37 -> GRCh38)
# SNPs from Globularity GWAS are in GRCh37, Ensembl phylogenetic context works for GRCh38 coordinates. So, convert all SNPs to GRCh37 first.
# Do this using Crossap from Ensembl, which works on Python.

```{r fetch_seqs_100bp}

for (i in unique(lead_snps$Uploaded_variation)) {
  #i=2
  #i="rs74594184"
  tmp_rsID = i
  tmp_loc = lead_snps[lead_snps$Uploaded_variation==i,]$Location[1]
  # extract the region covering +-100bp of the variant
  tmp_chr = strsplit(tmp_loc, split = ":")[[1]][1]
  tmp_pos = as.numeric(strsplit(strsplit(tmp_loc, split = ":")[[1]][2], split = "-")[[1]][1])
  tmp_seq_str = tmp_pos-100
  tmp_seq_end = tmp_pos+100
  
  server <- "http://rest.ensembl.org"
  ext <- paste0("/alignment/region/homo_sapiens/", tmp_chr, ":", tmp_seq_str, "-", 
                tmp_seq_end, "?species_set_group=primates")

  r <- GET(paste(server, ext, sep = ""), content_type("application/json"))
  #r <- GET(paste(server, ext, sep = ""), content_type(".xml"))
  #stop_for_status(r)
 
  # Some of the variants fall within regions that are missing in non-human ape genome assemblies. 
  # Get their alignment (with species that have DNAseq data from those regions) from the Ensembl 
  # browser manually.
  #result = class(try(toJSON(content(r)), silent = T))
  
  if (class(try(toJSON(content(r)), silent = T)) == "try-error") {
    print(paste("Manually check:", tmp_rsID))
    next
  } 
  else if (fromJSON(toJSON(content(r)))[[1]] == "no alignment available for this region") {
    print(paste("Manually check:", tmp_rsID))
    next
  } else {
  
    json_object = fromJSON(toJSON(content(r)))[[1]]
    species_tree = json_object$tree
  
  # Create a df to fill with data from Ensembl
  
  m = matrix(NA, nrow = length(json_object$alignments), ncol = 6)
  tmp_df = as.data.frame(m)
  colnames(tmp_df) = c("species", 
                       "chr", 
                       "seq_str", 
                       "seq_end",  
                       "strand", 
                       "seq")
  
    # loop over extant species and ancestors
  for (j in 1:length(json_object$alignments)) {
    #j=1
    
    tmp_df$species[j] = json_object$alignments[[j]]$species
    tmp_df$chr[j] = tmp_chr
    tmp_df$seq_str[j] = json_object$alignments[[j]]$start
    tmp_df$seq_end[j] = json_object$alignments[[j]]$end
    tmp_df$strand[j] = json_object$alignments[[j]]$strand
    tmp_df$seq[j] = json_object$alignments[[j]]$seq
    
  }

  write.table(tmp_df, file = paste0(outDir, "100bp/", tmp_rsID, ".tab"),
                row.names = FALSE, quote = FALSE, sep = "\t")

  write.table(species_tree, file = paste0(outDir, "100bp/", tmp_rsID, "_tree.txt"),
                row.names = FALSE, quote = FALSE)
  
  }
}

```

## +-10bp

```{r fetch_seqs_10bp}
for (i in unique(lead_snps$Uploaded_variation)) {
  #i=2
  tmp_rsID = i
  tmp_loc = lead_snps[lead_snps$Uploaded_variation==i,]$Location[1]
  # extract the region covering +-100bp of the variant
  tmp_chr = strsplit(tmp_loc, split = ":")[[1]][1]
  tmp_pos = as.numeric(strsplit(strsplit(tmp_loc, split = ":")[[1]][2], split = "-")[[1]][1])
  tmp_seq_str = tmp_pos-10
  tmp_seq_end = tmp_pos+10
  
  server <- "http://rest.ensembl.org"
  ext <- paste0("/alignment/region/homo_sapiens/", tmp_chr, ":", tmp_seq_str, "-", 
                tmp_seq_end, "?species_set_group=primates")

  r <- GET(paste(server, ext, sep = ""), content_type("application/json"))
  #r <- GET(paste(server, ext, sep = ""), content_type(".xml"))
  #stop_for_status(r)
 
   # Some of the variants fall within regions that are missing in non-human ape genome assemblies. 
  # Get their alignment (with species that have DNAseq data from those regions) from the Ensembl 
  # browser manually.
  #result = class(try(toJSON(content(r)), silent = T))
  
  if (class(try(toJSON(content(r)), silent = T)) == "try-error") {
    print(paste("Manually check:", tmp_rsID))
    next
  } 
  else if (fromJSON(toJSON(content(r)))[[1]] == "no alignment available for this region") {
    print(paste("Manually check:", tmp_rsID))
    next
  } else {
    
    json_object = fromJSON(toJSON(content(r)))[[1]]
    species_tree = json_object$tree
  
  m = matrix(NA, nrow = length(json_object$alignments), ncol = 6)
  tmp_df = as.data.frame(m)
  colnames(tmp_df) = c("species", 
                       "chr", 
                       "seq_str", 
                       "seq_end",  
                       "strand", 
                       "seq")
  
  # loop over extant species and ancestors
  for (j in 1:length(json_object$alignments)) {
    #j=1
    
    tmp_df$species[j] = json_object$alignments[[j]]$species
    tmp_df$chr[j] = tmp_chr
    tmp_df$seq_str[j] = json_object$alignments[[j]]$start
    tmp_df$seq_end[j] = json_object$alignments[[j]]$end
    tmp_df$strand[j] = json_object$alignments[[j]]$strand
    tmp_df$seq[j] = json_object$alignments[[j]]$seq
    
  }

  write.table(tmp_df, file = paste0(outDir, "10bp/", tmp_rsID, "_10bp.tab"),
                row.names = FALSE, quote = FALSE, sep = "\t")

  write.table(species_tree, file = paste0(outDir, "10bp/", tmp_rsID, "_10bp_tree.txt"),
                row.names = FALSE, quote = FALSE)
  
  }
}
```

# Plot trees

```{r}

foo <- function() {
  col <- "green"
  for (i in 1:2)
  axis(i, col = col, col.ticks = col, col.axis = col, las = 1)
  box(lty = "19")
}

for (i in dir(paste0(outDir, "10bp"), pattern = "tree.txt", full.names = T)) {
  
  tmp_rsID = strsplit(strsplit(i, split = "_")[[1]][4], split = "/")[[1]][3]
  x = read.tree(text = read.table(i, header = T)[1,1])
  
  pdf(paste0(outDir, "10bp/", tmp_rsID, "_10bp_tree.pdf"))
  plot(x); foo()
  dev.off()
}
#i=dir(paste0(outDir, "10bp"), pattern = "tree", full.names = T)[1]
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
